src/decode/xz.rs

```diff
@@ -223,31 +227,22 @@ where
         )));
     }
 
-    let mut tmpbuf: Vec<u8> = Vec::new();
-    let filters = block_header.filters;
-    for (i, filter) in filters.iter().enumerate() {
-        if i == 0 {
-            // TODO: use SubBufRead on input if packed_size is known?
-            let packed_size = decode_filter(count_input, &mut tmpbuf, filter)?;
-            if let Some(expected_packed_size) = block_header.packed_size {
-                if (packed_size as u64) != expected_packed_size {
-                    return Err(error::Error::XzError(format!(
-                        "Invalid compressed size: expected {} but got {}",
-                        expected_packed_size, packed_size
-                    )));
-                }
-            }
-        } else {
-            let mut newbuf: Vec<u8> = Vec::new();
-            decode_filter(
-                &mut io::BufReader::new(tmpbuf.as_slice()),
-                &mut newbuf,
-                filter,
-            )?;
-            // TODO: does this move or copy?
-            tmpbuf = newbuf;
+    let decompress_filters = block_header.filters.iter().rev().collect::<Vec<_>>();
+    let mut tmpbuf = Vec::with_capacity(block_header.unpacked_size.unwrap_or(1 << 12) as usize);
```

Comment:
- Using a default size of 4KB sounds like a good idea. It would be good to create a constant at the start of the function and document why this constant has been chosen, rather than sprinkle `1 << 12` several times in the function without explanation for the reader.
- As far as I understand it, the `block_header.unpacked_size` is the size after all the filters have been applied, so there could be a risk to over-allocate for intermediate layers. Not for delta though.
- In the future, it could be nice to avoid intermediate allocations and directly pass through the filters (so a delta filter wouldn't require storing all the intermediate output in memory), but that would require some re-architecting so that's be for another day.

Reply:
Good call! I changed the constant to 4_096 as well. I think I was thinking that using a shift would clearly communicate that the number is a clean power-of-two. In retrospect, I think was just too deep in the weeds of crafting decoders. A simple number will be clearer to most audiences, including myself, now that it's a couple months later.